File: pc_nodes_examples/python/led_service_client.py
#!/usr/bin/env python3
import sys
import os
import random

# Ensure the shared zenoh_ros package can be imported
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../shared_libraries/python')))

from zenoh_ros import ZenohNode, ZenohConfig
from zenoh_ros.custom_srvs import z_SetLEDColor

class LEDServiceClientNode(ZenohNode):
    def __init__(self, service_name: str = "set_led_color") -> None:
        super().__init__("led_client_python")

        self.get_logger().info("==================================================")
        self.get_logger().info("    Zenoh ROS 2 Service Client (Random RGB LED)   ")
        self.get_logger().info("==================================================")

        self.service_name = service_name
        self.client = self.z_create_client(z_SetLEDColor, self.service_name)
        self.get_logger().info("Service client initialized on '%s'...", self.service_name)

        self.current_led = 0
        self._first_time = True

        # Create ROS 2 timer triggering callback every 1000ms (1.0s)
        self.timer = self.z_create_timer(1000, self.timer_callback)

    def timer_callback(self) -> None:
        try:
            if self._first_time:
                self.get_logger().info("Resetting all 16 LEDs off on startup...")
                for i in range(16):
                    reset_req = z_SetLEDColor.Request()
                    reset_req.led_data.led_num = i
                    reset_req.led_data.r = 0
                    reset_req.led_data.g = 0
                    reset_req.led_data.b = 0
                    reset_req.led_data.brightness = 0
                    self.client.call_async(reset_req)
                self._first_time = False
                self.z_delay(200)

            led_num = self.current_led

            r = random.randint(0, 255)
            g = random.randint(0, 255)
            b = random.randint(0, 255)
            brightness = 60

            # Construct service request with nested led_data
            req = z_SetLEDColor.Request()
            req.led_data.led_num = led_num
            req.led_data.r = r
            req.led_data.g = g
            req.led_data.b = b
            req.led_data.brightness = brightness

            self.get_logger().info("[Service Request] Set LED #%d -> RGB(%d, %d, %d) Brightness=%d", led_num, r, g, b, brightness)

            # Call service asynchronously with ROS 2 Future pattern inside try/except block
            future = self.client.call_async(req)
            res = future.result(timeout_sec=3.0)

            if res and hasattr(res, 'success') and res.success:
                self.get_logger().info("[Service Response] Success=%s | Message='%s'\n", res.success, res.message)
                self.current_led = (self.current_led + 1) % 16
            else:
                self.get_logger().warn("[Service Response] Call returned empty or failed response!\n")
        except Exception as e:
            self.get_logger().warn("[Service Response] Service call timed out or failed: %s\n", e)


def main() -> None:
    host_ip = sys.argv[1] if len(sys.argv) > 1 else "192.168.4.1"
    cfg = ZenohConfig(host=host_ip, port=7447)
    if not ZenohNode.init(cfg):
        return

    node_instance = None
    try:
        node_instance = LEDServiceClientNode()
        node_instance.z_spin()
    except KeyboardInterrupt:
        if node_instance:
            node_instance.get_logger().info("Stopping client node...")
    except Exception as e:
        if node_instance:
            node_instance.get_logger().error(f"Critical Error: {e}")
    finally:
        ZenohNode.shutdown()
        if node_instance:
            node_instance.get_logger().info("Zenoh session cleanly closed.")

if __name__ == "__main__":
    main()

File: pc_nodes_examples/cpp/src/led_service_client.cpp
#include <zenoh_ros/ZenohRosPC.h>
#include <zenoh_ros/custom_srvs/z_SetLEDColor.h>
#include <cstdlib>
#include <ctime>

class LEDServiceClientNode : public ZenohNode {
public:
    LEDServiceClientNode(const char* service_name = "set_led_color")
        : ZenohNode("led_client_cpp"),
          service_name_(service_name),
          current_led_(0),
          first_time_(true) {

        ZLOG_INFO(this->get_logger(), "==================================================");
        ZLOG_INFO(this->get_logger(), "    Zenoh ROS 2 Service Client (Random RGB LED)   ");
        ZLOG_INFO(this->get_logger(), "==================================================");

        // Seed random number generator
        std::srand(std::time(nullptr));

        // Create service client for set_led_color service
        client_ = this->z_create_client<z_SetLEDColor>(service_name_);
        ZLOG_INFO(this->get_logger(), "Service client initialized on '%s'...", service_name_);

        // Create ROS 2 timer triggering callback every 1000ms (1.0s)
        timer_ = this->z_create_timer(1000, [this]() -> void {
            this->timer_callback();
        });
    }

private:
    const char* service_name_;
    ZenohClient<z_SetLEDColor>* client_ = nullptr;
    ZenohTimer* timer_ = nullptr;
    int current_led_;
    bool first_time_;

    void timer_callback() {
        if (!client_ || shutdown_requested) return;

        if (first_time_) {
            ZLOG_INFO(this->get_logger(), "Resetting all 16 LEDs off on startup...");
            for (int i = 0; i < 16; ++i) {
                if (shutdown_requested) return;
                z_SetLEDColor::Request reset_req;
                reset_req.led_data.led_num = i;
                reset_req.led_data.r = 0;
                reset_req.led_data.g = 0;
                reset_req.led_data.b = 0;
                reset_req.led_data.brightness = 0;

                z_SetLEDColor::Response reset_res;
                client_->call(reset_req, reset_res, 1000);
            }
            first_time_ = false;
            this->z_delay(200);
        }

        if (shutdown_requested) return;

        int led_num = current_led_;
        uint8_t r = std::rand() % 256;
        uint8_t g = std::rand() % 256;
        uint8_t b = std::rand() % 256;
        uint8_t brightness = 60;

        // Construct service request with nested led_data
        z_SetLEDColor::Request req;
        req.led_data.led_num = led_num;
        req.led_data.r = r;
        req.led_data.g = g;
        req.led_data.b = b;
        req.led_data.brightness = brightness;

        ZLOG_INFO(this->get_logger(),
                  "[Service Request] Set LED #%d -> RGB(%d, %d, %d) Brightness=%d",
                  led_num, r, g, b, brightness);

        z_SetLEDColor::Response res;
        bool success = client_->call(req, res, 3000);

        if (shutdown_requested) return;

        if (success && res.success) {
            ZLOG_INFO(this->get_logger(),
                      "[Service Response] Success=True | Message='%s'\n",
                      res.message.c_str());
            current_led_ = (current_led_ + 1) % 16;
        } else {
            ZLOG_WARN(this->get_logger(),
                      "[Service Response] Call returned empty or failed response!\n");
        }
    }
};

int main(int argc, char** argv) {
    ZenohConfig config;
    if (argc > 1) {
        config.communication_mode = ZenohConfig::ZENOH_COMM_WIFI;
        config.host = argv[1];
        config.port = 7447;
    } else {
        config.communication_mode = ZenohConfig::ZENOH_COMM_WIFI;
        config.host = "192.168.4.1";
        config.port = 7447;
    }

    if (!ZenohNode::init(config)) {
        return -1;
    }

    {
        LEDServiceClientNode node;
        node.z_spin();
    }

    return 0;
}
